cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBMessageTable.java

```diff
@@ -102,14 +106,49 @@ protected void delete(byte[] startKey, byte[] stopKey) throws IOException {
         writeBatch.delete(rowIterator.next().getKey());
       }
     }
-    levelDB.write(writeBatch, WRITE_OPTIONS);
+
+    try {
+      levelDB.write(writeBatch, WRITE_OPTIONS);
+    } catch (DBException ex) {
+      throw new IOException(ex);
+    }
   }
 
   @Override
   public void close() throws IOException {
     // no-op
   }
 
+  /**
+   * Delete messages of a {@link TopicId} that has exceeded the TTL
+   *
+   * @param topicMetadata {@link TopicMetadata}
+   * @param currentTime current timestamp
+   * @throws IOException error occurred while trying to delete a row in LevelDB
+   */
+  public void pruneMessages(TopicMetadata topicMetadata, long currentTime) throws IOException {
+    WriteBatch writeBatch = levelDB.createWriteBatch();
+    long ttlInMs = TimeUnit.SECONDS.toMillis(topicMetadata.getTTL());
+    byte[] startRow = MessagingUtils.toRowKeyPrefix(topicMetadata.getTopicId());
+    byte[] stopRow = Bytes.stopKeyForPrefix(startRow);
+
+    try (CloseableIterator<Map.Entry<byte[], byte[]>> rowIterator = new DBScanIterator(levelDB, startRow, stopRow)) {
+      while (rowIterator.hasNext()) {
```

Comment:
You can terminate once you see an entry with publish time after TTL to avoid scanning the whole table, since entries are sorted by time.